adapters/sovrn/sovrn.go

```diff
@@ -0,0 +1,272 @@
+package sovrn
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prebid/prebid-server/pbs"
+	"golang.org/x/net/context/ctxhttp"
+)
+
+type SovrnAdapter struct {
+	http *adapters.HTTPAdapter
+	URI  string
+}
+
+// Name - export adapter name */
+func (s *SovrnAdapter) Name() string {
+	return "sovrn"
+}
+
+// FamilyName used for cookies and such
+func (s *SovrnAdapter) FamilyName() string {
+	return "sovrn"
+}
+
+func (s *SovrnAdapter) SkipNoCookies() bool {
+	return false
+}
+
+// Call send bid requests to sovrn and receive responses
+func (s *SovrnAdapter) Call(ctx context.Context, req *pbs.PBSRequest, bidder *pbs.PBSBidder) (pbs.PBSBidSlice, error) {
+	supportedMediaTypes := []pbs.MediaType{pbs.MEDIA_TYPE_BANNER}
+	sReq, err := adapters.MakeOpenRTBGeneric(req, bidder, s.FamilyName(), supportedMediaTypes, true)
+
+	if err != nil {
+		return nil, err
+	}
+
+	sovrnReq := openrtb.BidRequest{
+		ID:   sReq.ID,
+		Imp:  sReq.Imp,
+		Site: sReq.Site,
+	}
+
+	// add tag ids to impressions
+	for i, unit := range bidder.AdUnits {
+		var params openrtb_ext.ExtImpSovrn
+		err = json.Unmarshal(unit.Params, &params)
+		if err != nil {
+			return nil, err
+		}
+		sovrnReq.Imp[i].Secure = sReq.Imp[i].Secure
+		sovrnReq.Imp[i].TagID = params.TagId
+	}
+
+	reqJSON, err := json.Marshal(sovrnReq)
+	if err != nil {
+		return nil, err
+	}
+
+	debug := &pbs.BidderDebug{
+		RequestURI: s.URI,
+	}
+
+	if req.IsDebug {
+		debug.RequestBody = string(reqJSON)
+		bidder.Debug = append(bidder.Debug, debug)
+	}
+
+	httpReq, _ := http.NewRequest("POST", s.URI, bytes.NewReader(reqJSON))
+	httpReq.Header.Set("Content-Type", "application/json")
+	if sReq.Device != nil {
+		addHeaderIfNonEmpty(httpReq.Header, "User-Agent", sReq.Device.UA)
+		addHeaderIfNonEmpty(httpReq.Header, "X-Forwarded-For", sReq.Device.IP)
+		addHeaderIfNonEmpty(httpReq.Header, "Accept-Language", sReq.Device.Language)
+		addHeaderIfNonEmpty(httpReq.Header, "DNT", strconv.Itoa(int(sReq.Device.DNT)))
+	}
+	if sReq.User != nil {
+		userID := strings.TrimSpace(sReq.User.BuyerUID)
+		if len(userID) > 0 {
+			httpReq.AddCookie(&http.Cookie{Name: "ljt_reader", Value: userID})
+		}
+	}
+	sResp, err := ctxhttp.Do(ctx, s.http.Client, httpReq)
+	defer sResp.Body.Close()
```

Comment:
Hate to say it, but... I don't think this is right either. See https://stackoverflow.com/a/33238755

I know that's talking about `client.Do`, and not `ctxhttp.Do`, but the source code of `ctxhttp.Do` is:

```
if client == nil {
		client = http.DefaultClient
	}
	resp, err := client.Do(req.WithContext(ctx))
        .... ctx stuff ...
        return resp, err
```

So... this should go _after_ the `if err != nil` check, but before anything else.